torchaudio/datasets/datasets.py

```diff
@@ -0,0 +1,419 @@
+import os
+import pickle
+import random
+from functools import partial, reduce
+from warnings import warn
+
+import torch
+
+import torchaudio
+
+
+class Cache:
+    """
+    Wrap a generator so that, whenever a new item is returned, it is saved to disk in a pickle.
+    """
+
+    def __init__(self, generator, location):
+        self.generator = generator
+        self.location = location
+
+        self._id = id(self)
+        self._cache = []
+        self._internal_index = 0
+
+    def __iter__(self):
+        self._internal_index = 0
+        return self
+
+    def __next__(self):
+        if self._internal_index < len(self):
+            item = self[self._internal_index]
+        else:
+            item = next(self.generator)
+
+            file = str(self._id) + "-" + str(len(self))
+            file = os.path.join(self.location, file)
+            self._cache.append(file)
+
+            os.makedirs(self.location, exist_ok=True)
+            with open(file, "wb") as file:
```

Comment:
pickling and files are a relatively strong assumption. lots of things can't be pickled. in terms of caching things on list for each input argument, you might end up with *a lot* of small files for something like images. if you want to write a read-only file cache (which is what this kind of is) you want to have a single file you append to. but random seek is not very good when it comes to files.

i'd say for now we stick to in-memory only caching

Reply:
Or if you want to serialize it, read out the entire generator into a list first and then save that, or append to an open file etc.

Reply:
For reference: [SpooledTemporaryFile](https://docs.python.org/3/library/tempfile.html#tempfile.SpooledTemporaryFile).

Reply:
pytorch also has `torch.save`.